File: tensorflow/contrib/cudnn_rnn/python/ops/packing_ops.py
# Copyright 2016 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Cudnn RNN operators."""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from tensorflow.python.ops import gen_packing_ops

def packed_sequence_alignment(
    sequence_lengths,
    name=None):
  """Docstring goes here

  Args:
    sequence_lengths: the sequence lengths.
    name: name of the operation.
  Returns:
    tuple of alignments and batch sizes
  Raises:
    ValueError: if rnn_mode or direction is invalid.
  """
  ret = gen_packing_ops.packed_sequence_alignment(
      sequence_lengths=sequence_lengths,
      name=name)
  return ret[0], ret[1]
  
def sequence_gather_scatter_indices(
    total_length,
    sequence_lengths,
    batch_order,
    name=None):
  """Docstring goes here

  Args:
    sequence_lengths: the sequence lengths.
    name: name of the operation.
  Returns:
    tuple of alignments and batch sizes
  Raises:
    ValueError: if rnn_mode or direction is invalid.
  """
  ret = gen_packing_ops.sequence_gather_scatter_indices(
      total_length=total_length,
      sequence_lengths=sequence_lengths,
      batch_order=batch_order,
      name=name)
  return ret

def pack_sequence(
    sequence,
    alignments,
    batch_sizes,
    name=None):
  """Docstring goes here

  Args:
    sequence_lengths: the sequence lengths.
    name: name of the operation.
  Returns:
    tuple of alignments and batch sizes
  Raises:
    ValueError: if rnn_mode or direction is invalid.
  """
  ret = gen_packing_ops.pack_sequence(
      sequence=sequence,
      alignments=alignments,
      batch_sizes=batch_sizes,
      name=name)
  return ret
  
def unpack_sequence(
    packed,
    alignments,
    batch_sizes,
    name=None):
  """Docstring goes here

  Args:
    sequence_lengths: the sequence lengths.
    name: name of the operation.
  Returns:
    tuple of alignments and batch sizes
  Raises:
    ValueError: if rnn_mode or direction is invalid.
  """
  ret = gen_packing_ops.unpack_sequence(
      packed=packed,
      alignments=alignments,
      batch_sizes=batch_sizes,
      name=name)
  return ret
  

File: tensorflow/core/ops/packing_ops.cc
/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/core/framework/common_shape_fns.h"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/shape_inference.h"
#include "tensorflow/core/lib/strings/strcat.h"

namespace tensorflow {
namespace {
using shape_inference::DimensionHandle;
using shape_inference::InferenceContext;
using shape_inference::ShapeHandle;

REGISTER_OP("PackedSequenceAlignment")
    .Input("sequence_lengths: T")
    .Output("alignments: T")
    .Output("batch_sizes: T")
    .Attr("T: {int8, int16, int32, int64}")
    .SetShapeFn([](InferenceContext* c) {
	  auto alignments_shape = c->MakeShape({
		  c->UnknownDim()
	  });
	  auto batch_sizes_shape = c->MakeShape({
		  c->UnknownDim()
	  });
      c->set_output(0, alignments_shape);
      c->set_output(1, batch_sizes_shape);
      return Status::OK();
    });
	
REGISTER_OP("SequenceGatherScatterIndices")
    .Input("total_length: T")
    .Input("sequence_lengths: T")
    .Input("batch_order: T")
    .Output("gather_scatter_indices: T")
    .Attr("T: {int8, int16, int32, int64}")
    .SetShapeFn([](InferenceContext* c) {
	  auto gather_scatter_indices_shape = c->MakeShape({
		  c->UnknownDim(), 2
	  });
      c->set_output(0, gather_scatter_indices_shape);
      return Status::OK();
    });
	
REGISTER_OP("PackSequence")
    .Input("sequence: T")
    .Input("alignments: Index")
    .Input("batch_sizes: Index")
    .Output("packed_sequence: T")
    .Attr("T: {int32, int64, float32, float64}")
    .Attr("Index: {int32, int64}")
    .SetShapeFn([](InferenceContext* c) {
	  auto sequence_shape = c->input(0);
	  auto packed_sequence_shape = c->MakeShape({
		  c->UnknownDim(),
		  c->Dim(sequence_shape, 2)
	  });
      c->set_output(0, packed_sequence_shape);
      return Status::OK();
    });
	
REGISTER_OP("UnpackSequence")
    .Input("packed: T")
    .Input("alignments: Index")
    .Input("batch_sizes: Index")
    .Output("sequence: T")
    .Attr("T: {int32, int64, float32, float64}")
    .Attr("Index: {int32, int64}")
    .SetShapeFn([](InferenceContext* c) {
	  auto packed_shape = c->input(0);
	  auto alignments_shape = c->input(1);
	  auto sequence_shape = c->MakeShape({
		  c->Dim(alignments_shape,0),
		  c->UnknownDim(),
		  c->Dim(packed_shape, 1)
	  });
      c->set_output(0, sequence_shape);
      return Status::OK();
    });

}
}  // namespace tensorflow
